pkg/ccl/changefeedccl/changefeed_dist.go

```diff
@@ -545,80 +547,129 @@ var rebalanceThreshold = settings.RegisterFloatSetting(
 ) ([]sql.SpanPartition, error) {
-	if len(p) <= 1 {
-		return p, nil
+	if len(partitions) <= 1 {
+		return partitions, nil
 	}
 
-	// Explode set of spans into set of ranges.
-	// TODO(yevgeniy): This might not be great if the tables are huge.
-	numRanges := 0
-	for i := range p {
-		spans, err := r.getRangesForSpans(ctx, p[i].Spans)
-		if err != nil {
-			return nil, err
+	// Create partition builder structs for the partitions array above.
+	var builders = make([]partitionWithBuilder, len(partitions))
+	var totalRanges int
+	for i, p := range partitions {
+		builders[i].p = p
+		builders[i].pIdx = i
+		for _, sp := range p.Spans {
+			builders[i].g.Add(sp)
+			rSpan, err := keys.SpanAddr(sp)
+			if err != nil {
+				return nil, err
+			}
+			for ri.Seek(ctx, rSpan.Key, kvcoord.Ascending); ; ri.Next(ctx) {
+				if !ri.Valid() {
+					return nil, ri.Error()
+				}
+				builders[i].numRanges += 1
+				totalRanges += 1
+				if !ri.NeedAnother(rSpan) {
+					break
+				}
+			}
 		}
-		p[i].Spans = spans
-		numRanges += len(spans)
 	}
 
 	// Sort descending based on the number of ranges.
-	sort.Slice(p, func(i, j int) bool {
-		return len(p[i].Spans) > len(p[j].Spans)
+	sort.Slice(builders, func(i, j int) bool {
+		return builders[i].numRanges > builders[j].numRanges
 	})
 
-	targetRanges := int((1 + sensitivity) * float64(numRanges) / float64(len(p)))
-
-	for i, j := 0, len(p)-1; i < j && len(p[i].Spans) > targetRanges && len(p[j].Spans) < targetRanges; {
-		from, to := i, j
-
-		// Figure out how many ranges we can move.
-		numToMove := len(p[from].Spans) - targetRanges
-		canMove := targetRanges - len(p[to].Spans)
-		if numToMove <= canMove {
-			i++
-		}
-		if canMove <= numToMove {
-			numToMove = canMove
-			j--
+	targetRanges := int(math.Ceil((1 + sensitivity) * float64(totalRanges) / float64(len(partitions))))
+	to := len(builders) - 1
+	from := 0
+	terminate := func() bool {
+		return from >= to
+	}
+	// In each iteration of the outer loop, check if `from` has too many ranges.
+	// If so, move them to other partitions which need more ranges
+	// starting from `to` and moving down.
+	// Otherwise, increment `from` and check again.
+	for ; !terminate(); from++ {
+		if builders[from].numRanges <= targetRanges {
+			continue
 		}
-		if numToMove == 0 {
-			break
+
+		// numToMove is the number of ranges which need to be moved out of `from`
+		// to other partitions.
+		numToMove := builders[from].numRanges - targetRanges
+		count := 0
+		needMore := func() bool {
+			return count < numToMove
 		}
+		// Iterate over all the spans in `from`.
+		for spanIdx := 0; !terminate() && needMore() && spanIdx < len(builders[from].p.Spans); spanIdx++ {
+			sp := builders[from].p.Spans[spanIdx]
+			rSpan, err := keys.SpanAddr(sp)
+			if err != nil {
+				return nil, err
+			}
+			// Iterate over the ranges in the current span.
+			for ri.Seek(ctx, rSpan.Key, kvcoord.Ascending); !terminate() && needMore(); ri.Next(ctx) {
```

Comment:
i wonder if there is a way to do it in sql.PartitionSpans...
Going to think a bit more about this... but for now, flushing out comments so far.